### PR TITLE
Adding support for Scalar Cryptography Extensions (Zkn -- Zbkx, Zkne, Zknd, Zknh) 

### PR DESCRIPTION
**Pull request by munailwaqar**
### Introduction
This PR adds support for Zbkx, Zkne, Zknd and Zknh extensions in the CVA6 core. It also adds the documentation and tests for these extensions. These changes have been tested with self-written single instruction tests and with the riscv-arch-tests. This PR will complete the Zkn - NIST Algorithm Suite extension.

### Implementation
**Zbkx Extension:**
Added support for the Zbkx instruction set. It essentially expands the Bitmanip extension with additional instructions useful in cryptography. These instructions are xperm8, xperm4.

**Zkne Extension:**
Added support for the Zkne instruction set. It essentially adds AES encryption support for scalar cryptography. These instructions are aes32esi, aes32esmi, aes64es, aes64esm, aes64ks1i, aes64ks2.

**Zknd Extension:**
Added support for the Zknd instruction set. It adds AES decryption support for scalar cryptography. These instructions are aes32dsi, aes32dsmi, aes64ds, aes64dsm, aes64im, aes64ks1i, aes64ks2.

*Note:*
The aes64ks1i and aes64ks2 instructions are present in both the Zknd and Zkne extensions.

**Zknh Extension:**
Added support for the Zknh instruction set. It adds the hash function instructions support for scalar cryptography. These instructions are sha256sig0, sha256sig1, sha256sum0, sha256sum1, sha512sig0h, sha512sig0l, sha512sig1h, sha512sig1l, sha512sum0r, sha512sum1r, sha512sig0, sha512sig1, sha512sum0, sha512sum1.

### Modifications
Updated the ALU and decoder to recognize and handle Zbkx instructions. For Zkne, Zknd & Zknh, the decoder will now select the AES unit as functional unit instead of the ALU.

The complete Zkn extension is added under the **ZKN** bit for ease of use. This configuration will also require the RVB (bitmanip) bit to be set.

***Note:*** 
The Zkn extension does not require the use of vectorial fpu.

**AES Functional Unit**
A new functional unit was created inside the execute stage that will handle all AES and Hashing instructions (Zkne, Zknd, Zknh).
A new package "aes_pkg" handles all AES functions such as sbox substitution, mix columns, etc.
aes_unit

*Table. 1 IO port connections with AES FU*
|Signals	|IO	|Description	|Connection	|Type|
|-----------|-----|------------------|-----------------|-------|
|clk_i	|in	|Subsystem Clock	|SUBSYSTEM	|logic|
|rst_ni	|in	|Asynchronous reset active low	|SUBSYSTEM	|logic|
|fu_data_i	|in	|Data needed to execute instruction	|Issue Stage	|logic |
|orig_instr_aes	|in	|Original instruction required bits	|Issue Stage	|logic [5:0]|
|result_o	|out	|Result of AES unit	|Execute Stage	|logic [CVA6Cfg.XLEN-1:0]|

**Documentation and Reference**
The official RISC-V Cryptography Extensions Volume I was followed to ensure alignment with ratification. The relevant documentation for Zbkx, Zkne, Zknd and Zknh instructions was also added.

### Verification
**Assembly Tests:**
The instructions were tested and verified with the K module of both 32 bit and 64 bit versions of the riscv-arch-tests to ensure proper functionality. These tests check for ISA compliance, edge cases and use assertions to ensure expected behavior.

### CODE REVIEW

**Comment by Gchauvon** [comment]
Hello, thank you for your contribution. 
We are running Spyglass on the cv32a65x configuration, withtout this Cryptography Extension. 
It detects new errors and warning. They are listed below, can you fix them ?

- [485]    UndrivenInTerm-ML                      Error      cva6/core/ex_stage.sv                                                                               378     10    Detected undriven input terminal cva6.ex_stage_i.aes_result[31:0]
- [181]    W123                                       Error       cva6/core/ex_stage.sv                                                                               378     10    Variable 'aes_result[31:0]' read but never set.[Hierarchy: ':cva6:ex_stage_i@ex_stage']
- [17F]    STARC05-2.1.5.3      NoVectorOnCondExpr    Warning    cva6/core/ex_stage.sv                                                                               377     2     Conditional expression (aes_valid_i) does not evaluate to a scalar [Hierarchy: ':cva6:ex_stage_i@ex_stage']
- [182]    W224                                       Warning     cva6/core/ex_stage.sv                                                                               377     5     Multi-bit expression 'aes_valid_i' found when one-bit expression expected[Hierarchy: ':cva6:ex_stage_i@ex_stage']

**Comment by munailwaqar** [comment]
@Gchauvon  I fixed the Spyglass errors but the CI run is still failing. However, the reason for the failure isnt is unclear. Could you please clarify what’s causing the failure or point me to any logs that show the issue?

Let me know if there's anything specific I should check.

File: core/include/cv64a6_imafdcv_sv39_config_pkg.sv [hunk 1]
```diff
@@ -35,7 +35,7 @@ package cva6_config_pkg;
   localparam CVA6ConfigAxiAddrWidth = 64;
   localparam CVA6ConfigAxiDataWidth = 64;
   localparam CVA6ConfigFetchUserEn = 0;
-  localparam CVA6ConfigFetchUserWidth = 1; // Just not to raise warnings
+  localparam CVA6ConfigFetchUserWidth = 1;  // Just not to raise warnings
```
Comment (Gchauvon) [inline core/include/cv64a6_imafdcv_sv39_config_pkg.sv:38]: I'm not sure this whitespace change is necessary. If it is from verible, keep it. Otherwise can you remove it.

**Review by Gchauvon — APPROVED** [review]
It looks good to me. 

The issue on CI is not related to this PR. We are currently working on resolving it.

**Comment by JeanRochCoulon** [comment]
Can you rebase ?

File: core/include/cv64a6_imafdcv_sv39_config_pkg.sv (hunk 1 above)
Comment (munailwaqar) [reply core/include/cv64a6_imafdcv_sv39_config_pkg.sv:38]: It is from verible.

**Comment by munailwaqar** [comment]
@JeanRochCoulon rebase completed

**Comment by jquevremont** [comment]
Hi @fatimasaleem, I'd like to update the requirement specification with the scalar crypto extensions supported by CVA6. Can you join tomorrow's CVA6 meeting?

**Comment by fatimasaleem** [comment]
Sure @jquevremont I will join.

**Comment by munailwaqar** [comment]
@isa084 Hi, 
The Github CI is running arch tests for cv64a6_imafdc_sv39_hpdcache_wb and cv64a6_imafdc_sv39_hpdcache configurations. According to the CI logs, verif/tests/testlist_riscv-arch-test-cv64a6_imafdc_sv39.yaml file is being used and the relevant tests are also being run (please check the screenshots).

Screenshot from 2025-04-11 17-18-27
Screenshot from 2025-04-11 17-19-17

**Comment by isa084** [comment]
@munailwaqar  

My apologies ! 

I overlooked the logs corresponding to the following two configs, where the testcases were indeed executed : 

 - execute-riscv64-tests (dv-riscv-arch-test, cv64a6_imafdc_sv39_hpdcache, veri-testharness)
 - execute-riscv64-tests (dv-riscv-arch-test, cv64a6_imafdc_sv39_hpdcache_wb, veri-testharness)

Now, given that these tests passed in CI, I am still trying to figure why they didn't go through in a local regression of a fresh clone of the branch https://github.com/10x-Engineers/cva6/tree/zbkx-branch

Does gcc version have anything to do with this ? The CI seems to use gcc-13.2, whereas the default toolchain version mentioned in util/toolchain-builder/README.md is gcc-13.1 .

Let me know of your thoughts on this.

**Comment by isa084** [comment]
Update : 

I was able to successfully reproduce the CI results for target cv64a6_imafdc_sv39_hpdcache_wb on my local machine using the following command : 
`DV_TARGET=cv64a6_imafdc_sv39_hpdcache_wb bash verif/regress/dv-riscv-arch-tests.sh` 

All tests in verif/tests/testlist_riscv-arch-test-cv64a6_imafdc_sv39.yaml passed as expected (including the scalar crypto extension testcases).

I have, however, discovered a possible bug in verif/sim/cva6.py which may sit at the root of my strange observations today.  It is independent of the changes proposed in this pull request. I will confirm with some further experiments, and then open a separate thread for it. I will leave this thread now, lest, I cause further confusion.

Thanks a lot for your patience.

**Comment by JeanRochCoulon** [comment]
Hello @munailwaqar
Thanks for your PR and patience
To be able to merge your PR,
* Can you add a license header in the 6 new files
* Rebase your repository, I have no right on 10x-Engineers Github domain
* What is the status of the possible bug ?
Thanks

**Comment by munailwaqar** [comment]
> Hello @munailwaqar Thanks for your PR and patience To be able to merge your PR,
> 
> * Can you add a license header in the 6 new files
> * Rebase your repository, I have no right on 10x-Engineers Github domain
> * What is the status of the possible bug ?

Hello @JeanRochCoulon. I have:
-  Added license headers to the 6 new files.
- Rebase completed.
- The bug https://github.com/openhwgroup/cva6/issues/2926 as reported by @isa084 affects this PR too. The issue is a simple string vs tuple problem (python detected  --> elif base in ("cv64a6_imafdc_sv39_wb"): ---- as a string instead of a tuple). I have fixed the bug and made it part of this PR too. (Thanks to @isa084 for the suggested solution).

**Comment by JeanRochCoulon** [comment]
Great @munailwaqar, I will merge
